### PR TITLE
Fix Dart plugin registrant interaction with 'flutter test'

### LINKED ISSUE #88794: Mac_staging plugin_test_ios keeps failing in staging
`Mac_staging plugin_test_ios` was recently enabled in staging, but keeps failing with the following log.

Related: https://github.com/flutter/flutter/issues/86802
Example build: https://ci.chromium.org/ui/p/flutter/builders/staging/Mac_staging%20plugin_test_ios/58/overview

Seems we need a higher version of xcode:
```
[plugin_test_ios] [STDOUT] stderr: [  +20 ms] Found "Xcode 12.0, Build version 12A7209". Xcode 12.0.1 or greater is required to develop for iOS.
[plugin_test_ios] [STDOUT] stdout: [   +3 ms] "flutter ios" took 5,064ms.
[plugin_test_ios] [STDOUT] stderr: [   +3 ms] Encountered error while building for device.
[plugin_test_ios] [STDOUT] stderr: [        ] 
[plugin_test_ios] [STDOUT] stderr:            #0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
[plugin_test_ios] [STDOUT] stderr:            #1      _BuildIOSSubCommand.runCommand (package:flutter_tools/src/commands/build_ios.dart:277:7)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #2      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1125:27)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #3      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #4      CommandRunner.runCommand (package:args/command_runner.dart:209:13)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #5      FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:288:9)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #6      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #7      FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:236:5)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #8      run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:62:9)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #9      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
[plugin_test_ios] [STDOUT] stderr:            #10     main (package:flutter_tools/executable.dart:92:3)
[plugin_test_ios] [STDOUT] stderr:            <asynchronous suspension>
```

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
Building an application for a desktop platform that transitively included any Dart-based plugins (such as path_provider) broke `flutter test`, because its compilation was overriding the provided main (in this case, the test main) with `generated_main.dart` if it was present. This PR:
- Changes the `flutter test` compilation path to update `generated_main.dart`, so that the tests will work, and will include any registered Dart plugins.
- Makes using `generated_main.dart` during recompile opt-in, to try to reduce the chance of a similar bug happening with other codepaths in the future.

Fixes https://github.com/flutter/flutter/issues/88794

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by blasten — COMMENTED** [review]
got it. Changes LGTM

**Comment by stuartmorgan** [comment]
@blasten I added a unit test, and the other tests seems happy, so this is ready for review.

File: packages/flutter_tools/test/general.shard/test/test_compiler_test.dart [hunk 1]
```diff
@@ -113,6 +115,57 @@ void main() {
...
+
+    final File generatedMain = fileSystem
+      .directory('.dart_tool')
+      .childDirectory('flutter_build')
+      .childFile('generated_main.dart');
+
+    expect(generatedMain, exists);
+    expect(
+      generatedMain.readAsStringSync(),
+      contains("import 'test/foo.dart' as entrypoint;")
```
Comment (blasten) [inline packages/flutter_tools/test/general.shard/test/test_compiler_test.dart:161]: what if someone changes the generated code to `// import 'test/foo.dart' as entrypoint;`?  
could we ensure the entire output?

**Review by blasten — APPROVED** [review]
LGTM

**Comment by fertrig** [comment]
Will this PR also fix `flutter run`? 

We have an issue where our tool runs `flutter build bundle`, which creates the `generated_main.dart` file, which imports a main file generated by the tool. Then, when the user runs `flutter run` with some other main target, the old `generated_main.file` is still used.

**Comment by stuartmorgan** [comment]
> Will this PR also fix `flutter run`?

I doubt it; I didn't intentionally change anything for the `run` path.

> We have an issue where our tool runs `flutter build bundle`, which creates the `generated_main.dart` file, which imports a main file generated by the tool. Then, when the user runs `flutter run` with some other main target, the old `generated_main.file` is still used.

Please file an issue with detailed repro steps and @-mention me; the `test` version is the only issue I was aware of.

**Comment by stuartmorgan** [comment]
`Google testing` isn't updating for some reason, but I've checked that the associated CL did pass its tests, so I'm going to land.

**Pull request merged**